flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -74,88 +76,233 @@
-			final State<T> beginningState;
+			states.add(beginningState);
+			usedNames.add(beginningState.getName());
+
+			return beginningState;
+		}
+
+		private State<T> convertToTimesState(State<T> sinkState, State<T> sourceState, int times) {
+			convertToSingletonState(sinkState, sourceState);
+			for (int i = 0; i < times - 1; i++) {
+				sinkState = sourceState;
+				sourceState = new State<>(currentPattern.getName(), State.StateType.Normal);
+				states.add(sourceState);
+				convertToSingletonState(sinkState, sourceState);
+			}
+			return sourceState;
+		}
+
+		@SuppressWarnings("unchecked")
+		private void convertToSingletonState(
+			final State<T> sinkState,
+			final State<T> sourceState) {
+
+			final FilterFunction<T> currentFilterFunction = (FilterFunction<T>) currentPattern.getFilterFunction();
+			final FilterFunction<T> trueFunction = FilterFunctions.trueFunction();
+			sourceState.addTake(sinkState, currentFilterFunction);
+
+			final State<T> ignoreState;
+			if (currentPattern.getQuantifier() == Quantifier.OPTIONAL) {
+				sourceState.addProceed(sinkState, trueFunction);
+				ignoreState = new State<>(currentPattern.getName(), State.StateType.Normal);
 
-			if (states.containsKey(BEGINNING_STATE_NAME)) {
-				beginningState = states.get(BEGINNING_STATE_NAME);
+				ignoreState.addTake(sinkState, currentFilterFunction);
+				states.add(ignoreState);
 			} else {
-				beginningState = new State<>(BEGINNING_STATE_NAME, State.StateType.Start);
-				states.put(BEGINNING_STATE_NAME, beginningState);
+				ignoreState = sourceState;
+			}
+
+			if (currentPattern instanceof FollowedByPattern) {
+				sourceState.addIgnore(ignoreState, trueFunction);
+			}
+		}
+
+		@SuppressWarnings("unchecked")
+		private State<T> createFirstMandatoryStateOfLoop(
+			final State<T> sinkState,
+			final State.StateType stateType) {
+
+			final FilterFunction<T> currentFilterFunction = (FilterFunction<T>) currentPattern.getFilterFunction();
+			final State<T> firstState = new State<>(currentPattern.getName(), stateType);
+
+			firstState.addTake(sinkState, currentFilterFunction);
+			if (currentPattern instanceof FollowedByPattern) {
+				if (currentPattern.getQuantifier() == Quantifier.ONE_OR_MORE_COMBINATIONS) {
+					firstState.addIgnore(FilterFunctions.<T>trueFunction());
+				} else {
+					firstState.addIgnore(new NotFilterFunction<>(currentFilterFunction));
+				}
 			}
+			return firstState;
+		}
 
-			beginningState.addStateTransition(new StateTransition<T>(
-				StateTransitionAction.TAKE,
-				currentState,
-				(FilterFunction<T>) currentPattern.getFilterFunction()
-			));
+		@SuppressWarnings("unchecked")
+		private void convertToLooping(
+			final State<T> sinkState,
+			final State<T> sourceState) {
 
-			return new NFAFactoryImpl<T>(inputTypeSerializer, windowTime, new HashSet<>(states.values()), timeoutHandling);
+			final FilterFunction<T> filterFunction = (FilterFunction<T>) currentPattern.getFilterFunction();
```

Comment:
You can add a check: `Preconditions.checkArgument(currentPattern.getQuantifier().isLooping());`